### PR TITLE
implement except node

### PR DESCRIPTION
**Pull request by yangzhg**
#2845
implement except node,
support  statement like 
``` 
select a from t1 except select b from t2
```

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/SetOperationStmt.java [hunk 1]
```diff
@@ -649,6 +648,13 @@ public void analyze(Analyzer parent) throws AnalysisException, UserException {
             if (isAnalyzed()) {
                 return;
             }
+            if (queryStmt instanceof SelectStmt && ((SelectStmt) queryStmt).fromClause_.isEmpty()) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/SetOperationStmt.java]: What this for? Could you add comment to explain this rewrite?

File: be/src/exec/except_node.cpp [hunk 2]
```diff
@@ -56,4 +40,157 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+            // if this hash table is probed rebuild it
+            std::unique_ptr<HashTable> temp_tbl(
+                    new HashTable(_child_expr_lists[0], _child_expr_lists[i], _build_tuple_size,
+                                  true, _find_nulls, id(), mem_tracker(), 1024));
+            _hash_tbl_iterator = _hash_tbl->begin();
+            uint32_t previous_hash = -1;
+            while (_hash_tbl_iterator.has_next()) {
+                if (previous_hash != _hash_tbl_iterator.get_hash()) {
+                    previous_hash = _hash_tbl_iterator.get_hash();
+                    if (!_hash_tbl_iterator.matched()) {
```
Comment (wutiangan) [inline be/src/exec/except_node.cpp]: why do you use  'if (!_hash_tbl_iterator.matched())'?
may be you should use ' if (_hash_tbl_iterator.matched())'
matched means exceptNode

File: be/src/exec/except_node.cpp [hunk 3]
```diff
@@ -56,4 +40,157 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+Status ExceptNode::get_next(RuntimeState* state, RowBatch* out_batch, bool* eos) {
+    RETURN_IF_ERROR(exec_debug_action(TExecNodePhase::GETNEXT));
+    RETURN_IF_CANCELLED(state);
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    *eos = true;
+    if (reached_limit()) {
+        return Status::OK();
+    }
+    uint32_t previous_hash = -1;
+    while (_hash_tbl_iterator.has_next()) {
```
Comment (wutiangan) [inline be/src/exec/except_node.cpp]: How about if the hash value is - 1？

File: be/src/exec/except_node.cpp [hunk 4]
```diff
@@ -56,4 +40,157 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+        if (previous_hash != _hash_tbl_iterator.get_hash()) {
+            previous_hash = _hash_tbl_iterator.get_hash();
+            if (!_hash_tbl_iterator.matched()) {
+                int row_idx = out_batch->add_row();
+                TupleRow* out_row = out_batch->get_row(row_idx);
+                uint8_t* out_ptr = reinterpret_cast<uint8_t*>(out_row);
+                memcpy(out_ptr, _hash_tbl_iterator.get_row(), _build_tuple_row_size);
+                out_batch->commit_last_row();
+            }
+        }
```
Comment (wutiangan) [inline be/src/exec/except_node.cpp]: If the two hash values are equal, it does not mean that the two values are equal. You need to compare the original values.
There may be a problem with the else's logic

File: be/src/exec/except_node.cpp (hunk 2 above)
Comment (yangzhg) [reply be/src/exec/except_node.cpp]: a excepte b , use a to build a hash table and use b to probe
 the result is  element in a and never mached

File: be/src/exec/except_node.cpp (hunk 3 above)
Comment (yangzhg) [reply be/src/exec/except_node.cpp]: because hash value  is a unsigned int

**Review by morningman — DISMISSED** [review]
LGTM

File: be/src/exec/exec_node.cpp [hunk 5]
```diff
@@ -446,9 +446,9 @@ Status ExecNode::create_node(RuntimeState* state, ObjectPool* pool, const TPlanN
         *node = pool->add(new IntersectNode(pool, tnode, descs));
         return Status::OK();
 
-    // case TPlanNodeType::EXCEPT_NODE:
-    //     *node = pool->add(new ExceptNode(pool, tnode, descs));
-    //     return Status::OK();
```
Comment (kangkaisen) [inline be/src/exec/exec_node.cpp]: Remove this line.

File: be/src/exec/except_node.cpp [hunk 6]
```diff
@@ -56,4 +41,190 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+        *eos = !_hash_tbl_iterator.has_next() || reached_limit();
+        if (out_batch->is_full() || out_batch->at_resource_limit() || *eos) {
+            return Status::OK();
+        }
+    }
+    return Status::OK();
+}
+
+bool ExceptNode::equals(TupleRow* row, TupleRow* other) {
+    if (row == nullptr && other == nullptr) {
```
Comment (kangkaisen) [inline be/src/exec/except_node.cpp]: This if statement could remove ?

File: be/src/exec/except_node.cpp [hunk 7]
```diff
@@ -56,4 +41,190 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+        return false;
+    }
+    if (row == nullptr || other == nullptr) {
+        return false;
+    }
+    for (int i = 0; i < _child_expr_lists[0].size(); ++i) {
+        void* val_row = _child_expr_lists[0][i]->get_value(row);
+        void* val_other = _child_expr_lists[0][i]->get_value(other);
+        if (val_row == nullptr && val_other == nullptr) {
+            continue;
```
[the remark this answers is not in the document]
Comment (yangzhg) [reply be/src/exec/except_node.cpp]: for nullable value when is null

File: be/src/exec/exec_node.cpp (hunk 5 above)
Comment (yangzhg) [reply be/src/exec/exec_node.cpp]: comment in wrong place ？

File: be/src/exec/except_node.cpp (hunk 6 above)
Comment (yangzhg) [reply be/src/exec/except_node.cpp]: this for a safe guard

Comment (kangkaisen) [reply be/src/exec/except_node.cpp]: I don't see any meaning.

File: be/src/exec/except_node.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/exec/except_node.cpp]: If `_child_expr_lists[0].size` is 1. `ExceptNode::equals` will return true. which is your expectation.  which should be conflicting with `else if (val_row == nullptr || val_other == nullptr) {
            return false;`

[the remark this answers is not in the document]
Comment (yangzhg) [reply be/src/exec/except_node.cpp]: if table a is 
```
c1
---
null
```

table b is 
```
c2
---
null
```
if caluse is 
`select c1 from a except select c2 from b `
will hit `val_row == nullptr && val_other == nullptr`

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**